### ISSUE TITLE
[explicit-function-return-type] Suggestion: disallow inferred types

### ISSUE THREAD

**Issue by OliverJAsh** [typescript-eslint/typescript-eslint#3379]
(I have erased the template because this is a suggestion, not a bug.)

Consider the following code:

```TS
// Errors, good ✅
const arrowFn1 = () => 'test';

// Doesn't error, good ✅
type FuncType = () => string;
const arrowFn2: FuncType = () => 'test';

type State = {
    foo: string;
    bar: string;
};

type Action = { tag: 'a' } | { tag: 'b' };

declare const matchAction: <T>(
    action: Action,
    matchers: Record<Action['tag'], () => T>,
) => T;

const reducer1 = (state: State, action: Action): State =>
    matchAction(action, {
        // Doesn't error, bad ❌
        a: () => ({
            ...state,
            foo: 'abc',
        }),
        b: () => ({
            ...state,
            foo: 'abc',
        }),
    });
```

I understand we don't get an error here (see comment in `reducer`) because `allowTypedFunctionExpressions` defaults to `true`.

I believe I actually want `allowTypedFunctionExpressions`. However, the behaviour I want is slightly more nuanced.

I want to allow typed function expressions but only if their return type is a named type—not an inferred type as it is in the case of the functions we pass to `matchAction`.

If we had a way to enable this behaviour, this would be allowed:

```ts
const reducer2 = (state: State, action: Action): State =>
    matchAction<State>(action, {
        a: () => ({
            ...state,
            foo: 'abc',
        }),
        b: () => ({
            ...state,
            foo: 'abc',
        }),
    });
```

For context, I'm trying to ensure that everywhere an object literal is defined, TypeScript understands which named type it relates to, so that tools such as "go to definition" / "find references" / "rename" on object properties work properly and to avoid bugs. Here are some examples of what I mean by that:

In my `reducer1` example above, if you try to rename the property `foo` inside one of the object literals, it won't update the name of the corresponding property in the `State` type definition. This is because TypeScript does not understand that this object literal relates to the `State` type.

https://user-images.githubusercontent.com/921609/117952197-d76a9d00-b30c-11eb-8b4c-a8c48161e4c6.mov

Imagine we rename `foo` in `State` to something like `baz`. We might forget to update the corresponding `foo` properties in our reducer, and TypeScript won't warn us because `foo` will be considered an excess property. (TypeScript only checks for excess properties if the object has an annotation, but in this case it does not because the return type is inferred.)

https://user-images.githubusercontent.com/921609/117952232-df2a4180-b30c-11eb-937c-b1cc6ae3e98c.mov

If we annotate the function return type we can resolve both of these issues.

**Comment by OliverJAsh** [comment]
Actually this problem isn't limited to return types. For example:

```ts
type State = {
    foo: string;
    bar: string;
};

type Action = { tag: 'a' } | { tag: 'b' };

const reducer1 = (state: State, action: Action): State => {
    const newState = {
        ...state,
        foo: 'abc',
    };
    return newState;
};

```

Maybe it makes sense to have a rule that requires type annotations for object literals, regardless of where they're used—in return types or as expressions for variable declarations.

**Comment by bradzacher** [comment]
Yeah the issue here is that:
1) your function is generically unconstrained, so TS just infers whatever object literal type it can.
2) typescript does not always do "exact object property" checks - it allows superset assignments in certain conditions (like in a return position). (i.e. in certain cases types are considered covariant [I think I got that right...? I always mix the variance terms up])
3) typescript does not "reverse infer" the generic type based on return type context.

Simplifying your example a little
```TS
interface State {
  foo: string;
  bar: string;
}

declare function matchAction<T>( // T here is unconstrained
    matchers: T,
): T;
```

```TS
// #1
matchAction({           // T = { foo: string, bar: string }
  ...state,
  foo: 'abc',
});

// #2
matchAction<State>({    // T = State
  ...state,
  foo: 'abc',
})

// #3
function reducer(): State {
  return matchAction({  // T = { foo: string, bar: string }
    ...state,           // T is assignable to State, so the return is valid
    foo: 'string',
  });
}
```

After using TS to auto-rename `State['foo']` to `State['foofoo']`

```TS
// #1
matchAction({           // T = { foofoo: string, bar: string, foo: string }
  ...state,
  foo: 'abc',
})

// #2
matchAction<State>({    // T = State
  ...state,
  foofoo: 'abc',        // the auto-rename will actually rename the property correctly here
})

// #3
function reducer(): State {
  return matchAction({  // T = { foofoo: string, bar: string, foo: string }
    ...state,           // T is still (loosely) assignable to State, so the return is valid
    foo: 'string',
  });
}
```

---

I wish there was a way to tell TypeScript to treat an object type invariantly.

[Flow does this via it's "exact object types"](https://flow.org/try/#0C4TwDgpgBAysCGxoF4oG8A+AoKUBmA9gQFxQDOwATgJYB2A5gDQ5QBG8lpFNDzGAvgG4sWACYQAxgBsO0PAFdaE4NQK0oAW0QSAFgEFlq2gB4AKgD4AFFuC6IlMqVMBKJ8KwS1FcgiSk4iCia2vqGapZo+ESkAOQxjGwcsfGJAF7JUPzOgkA)
```flow
type State = {|
  foo: string,
  bar: string,
|};

declare function matchAction<T>(matchers: T): T;

const state: State = matchAction({ foo: '', bar: '', baz: '' });
// Cannot assign `matchAction(...)` to `state` because property `baz` is missing in `State` [1] but exists in object literal [2].
```

(and fun fact flow has a compiler option to treat object types as "exact by default", so the code instead would be:
```flow
type State = { foo: string, bar: string };
type StateLoose = { foo: string, bar: string, ... }; // ... signals "loose" or "inexact" object and allows extra properties
```

**Comment by bradzacher** [comment]
> Maybe it makes sense to have a rule that requires type annotations for object literals, regardless of where they're used

This could be a decent enough proposal for a new rule, but it's really difficult to "fix" the problem.

For example - [doing an explicit `as` assertion can actually be really unsafe in TS](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=5&pc=45#code/MYewdgzgLgBAHjAvDA3gQwFwwOTYL4DcAUEaJLAJ5IwAUCaEqmM0ATgJZgDmeAlATAD0gmAFs0AawCmjCFMhTS4aDABe1OjAZMsbTlwA0MAEa6oHbnwHCYAdzTtYtgBZpYAWncwoz9oz9g2LBoYDBSrKwgrACEAPxxcUrkWtSqAHTGaQA28lw+1iLGWSC2jACuAA5asKxlYFDsolJAA):
```TS
const x = {a: ''};

const y = (x as {a: string}); // makes sense
const z = (x as {a: string, b: string}); // wait what -- this isn't an error!?!?!?
const a = z.b.length; // blows up at runtime
```

[Even if you annotate a type via a variable declaration annotation - you still have the same covariant behaviour](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=3&pc=39#code/MYewdgzgLgBAHgLhgbwIZOgJwJZgOYA0MARhlDvgL4wC8K6MA5I0aU45QNwBQ3oksAF5I0ZCnmp04nGAHpZMMCBgBTTJhCYAhEA):
```ts
const x: {a: string, b: string} = {a: '', b: ''};

const z: {a: string} = x; // no error!
```

**Comment by OliverJAsh** [comment]
Thanks for your reply!

> This could be a decent enough proposal for a new rule, but it's really difficult to "fix" the problem.

I would be happy to open a new issue for this if you think it makes sense to do so.

> [doing an explicit `as` assertion can actually be really unsafe in TS](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=5&pc=45#code/MYewdgzgLgBAHjAvDA3gQwFwwOTYL4DcAUEaJLAJ5IwAUCaEqmM0ATgJZgDmeAlATAD0gmAFs0AawCmjCFMhTS4aDABe1OjAZMsbTlwA0MAEa6oHbnwHCYAdzTtYtgBZpYAWncwoz9oz9g2LBoYDBSrKwgrACEAPxxcUrkWtSqAHTGaQA28lw+1iLGWSC2jACuAA5asKxlYFDsolJAA)

I think any assertion in TS is unsafe? I'm not concerned about _assertions_ causing problems, rather I'm only interested in _type annotations_.

> Even if you annotate a type via a variable declaration annotation - [you still have the same covariant behaviour](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=3&pc=39#code/MYewdgzgLgBAHgLhgbwIZOgJwJZgOYA0MARhlDvgL4wC8K6MA5I0aU45QNwBQ3oksAF5I0ZCnmp04nGAHpZMMCBgBTTJhCYAhEA):

I understand this is because of structural typing. I'm not sure this relates to the problems I described in my original post—at least solving this is not a goal of the lint rule I had in mind. I agree exact types would be nice, but I think we can solve the problems I mentioned if we just make sure every object literal has a type annotation. 🤔

**Comment by bradzacher** [comment]
I definitely agree with and understand the premise of your issue.
I don't think this is something TS could ever truly support due to the ambiguity so it's probably something that a linter could push the user towards.

But I think this should probably be a separate rule altogether rather than part of `explicit-function-return-type`.
You'd ideally want each and every object literal to have an annotated type.

Though it's worth noting that some cases are *really* hard to detect if they have a type for them or not.

```ts
const x = { a: 1 };         // trivial to detect
const y = () => ({ a: 1 }); // trivial to detect

declare function foo(arg: { a: 1 }): void;
foo({ a: 1 }); // harder to detect, but not that hard

declare function bar(arg: () => { a: 1 }): void;
bar(() => ({ a: 1 })); // harder to detect still, requires very special logic

declare function baz<T>(arg: T): void;
baz({ a: 1 }); // harder to detect

function foo2(): { a: 1 } {
  return { a: 1 }; // trivial to detect
}

const bar2: () => { a: 1 } = () => ({ a: 1 }); // hard to detect

function baz2(): { a: 1 } {
  return baz({ a: 1 }); // really hard to detect
}
```

If you go with a pure AST approach, then things can get a bit weird and are hard to detect
```ts
const x = { a: 1 };         // trivial to detect
const y = () => ({ a: 1 }); // trivial to detect

const bar2: () => { a: 1 } = () => ({ a: 1 }); // hard to detect

declare function baz<T>(arg: T): void;
baz({ a: 1 }); // impossible to detect

function baz2(): { a: 1 } {
  return baz({ a: 1 }); // impossible to detect
}
```

Not saying it's not doable or not worth doing - just trying to highlight that it's a really hard problem™

**Comment by OliverJAsh** [comment]
Yeah I can see that it's a difficult problem. 🤔 

Seeing as this no longer relates to the `explicit-function-return-type` lint rule, do you want me to close this one and open a new issue?

**Comment by bradzacher** [comment]
we can keep this issue - just update the title / body as appropriate :)

**Comment by OliverJAsh** [comment]
I hope you don't mind but I decided to open a new issue just to preserve the discussion here—the discussion could be confusing to future readers if we substantially change the title/body.

I'll close this one in favour of https://github.com/typescript-eslint/typescript-eslint/issues/3408.
